8284490: Remove finalizer method in java.security.jgss

XueleiFan opened the pull request:
Please review the update to remove finalizer method in the java.security.jgss module. It is one of the efforts to clean up the use of finalizer method in JDK.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8284490](https://bugs.openjdk.java.net/browse/JDK-8284490): Remove finalizer method in java.security.jgss


### Reviewers
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to 23072ef779c0400dfae60d1e85c7bd656d92ec04
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**) ⚠️ Review applies to aed8cd328f666f64aa485531d4128614495e741f
 * [Weijun Wang](https://openjdk.java.net/census#weijun) (@wangweij - **Reviewer**) ⚠️ Review applies to d554c7245e2ed72576139ed88f35f127e4232997


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8136/head:pull/8136` \
`$ git checkout pull/8136`

Update a local copy of the PR: \
`$ git checkout pull/8136` \
`$ git pull https://git.openjdk.java.net/jdk pull/8136/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8136`

View PR using the GUI difftool: \
`$ git pr show -t 8136`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8136.diff">https://git.openjdk.java.net/jdk/pull/8136.diff</a>

</details>

dfuch commented on src/java.security.jgss/share/classes/sun/security/jgss/wrapper/GSSCredElement.java:
```diff
@@ -69,6 +70,8 @@ void doServicePermCheck() throws GSSException {
         cStub = GSSLibStub.getInstance(mech);
         usage = GSSCredential.INITIATE_ONLY;
         name = srcName;
+
+        Cleaner.create().register(this, this::dispose);
```
This will create a memory leak: `this` will never be garbage collected if it's referenced by the cleaner action.

RogerRiggs replied:
The cleaner gets triggered when the instance is unreachable, no values from that instance can be modified or referenced.
Any information needed for the cleanup must be in a separate object.
In this case, that is pCred and cStub. 
Typically, new nested class is defined that holds the pCred and cStub and has a run() method that 
calls `cStub.releaseCred(pCred)`.
But the package access to non-final `pCred` raises a flag about when its mutated.
If its mutated after the cleaner is created, the cleaner will release the wrong cred.

If the GSSCredElement is itself just a holder of the credentials, then perhaps the cleaner should be triggered on the instance that is referring to the GSSCredElement being unreachable.

XueleiFan replied:
Yes.  I'm working on an update, and more code changes (including immutable objects) may be placed.  Thank you @dfuch and @RogerRiggs for the good catch.  I totally forgot about the object reference problems while using Cleaner.

valeriepeng replied:
Hmm, the earlier JCE change would also needs to be updated as it calls a cleanup method on the to-be-cleaned object.

XueleiFan replied:
> Hmm, the earlier JCE change would also needs to be updated as it calls a cleanup method on the to-be-cleaned object.

Yes, I will check the cleaner used in the security components and make sure there is object reference problems.

XueleiFan replied:
> I'm not sure if it's possible to write a test on the cleanup, but at least we can temporarily add a `println` line there and see if it ever gets called.

I did not find a way to test cleanup yet.  Yes, a temporary 'println" is what I can use for now.

RogerRiggs reviewed (approved):
LGTM

wangweij commented:
I see a problem. In `NativeGSSContext`, you only assign `cleanable` for one special ctor. In fact, the other 2 ctors (one for initiator and one for acceptor) are more useful. `pContext` is initialized to 0 in these 2 ctors and it's only assigned inside the native code after the context is established. There is no chance to dispose them anymore.

I'm not sure what the best way to do this, but maybe you can put `stub` and `pContext` is a separate object and always try to clean it even if `pContext` is not assigned at the beginning.

wangweij commented:
At least for native JGSS, all those native functions that release objects can print out debug messages if `-Dsun.security.nativegss.debug=true`. You can see whether they are called without adding extra `println` calls.

wangweij commented:
@bchristi-git showed me this test. You might try it on other objects:
```
import org.ietf.jgss.GSSManager;
import org.ietf.jgss.GSSName;

import java.util.WeakHashMap;

public class A1 {
    private static WeakHashMap whm = new WeakHashMap();
    public static void main(String[] args) throws Exception {
        System.setProperty("sun.security.nativegss.debug", "true");
        System.setProperty("sun.security.jgss.native", "true");
        GSSName e = GSSManager.getInstance().createName("u1", GSSName.NT_USER_NAME);
        whm.put(e, null);
        e = null;
        for (int i = 0; i < 10; i++) {
            System.gc();
            Thread.sleep(100);
        }
        if (whm.size() > 0) {
            throw new RuntimeException("GSSElementName still reachable");
        }
    }
}
```

bchristi-git commented:
Thanks, @wangweij . Wherever practical, it would be good to include tests to confirm correct conversions from finalizer to Cleaner -- bugs can be subtle, and hard to spot.

XueleiFan commented:
>  You might try it on other objects:

Nice idea to test object collection.  I added two test cases.

> The `NativeGSSContext` code still needs to be fixed.

Could you have more details?  I did not catch the comment about NativeGSSContext.

wangweij commented:
> Could you have more details? I did not catch the comment about NativeGSSContext.

When `NativeGSSContext(GSSNameElement peer, GSSCredElement myCred, int time, GSSLibStub stub)` is called, `pContext` is 0 and you haven't registered the cleaner. Later, when `initSecContext()` is called, it calls into `cStub.initContext()` and this native method will set `pContext` if a context is established. Since you haven't registered the cleaner in the ctor, this native context will not be released at the end.

So one solution is to add a `setNativeContext(long pContext)` method and when this method is called you register the cleaner. Now, inside the native method, instead of setting the `pContext` field directly you can call this setter method.

dfuch commented on src/java.security.jgss/share/classes/sun/security/jgss/wrapper/GSSNameElement.java:
```diff
@@ -185,7 +185,7 @@ private GSSNameElement() {
                                 printableType);
 
-        Cleaner.create().register(this, this::dispose);
+        cleanable = Krb5Util.cleaner.register(this, disposerFor(stub, pName));
```
I wonder if this line should be moved just after line 159?

dfuch commented on src/java.security.jgss/share/classes/sun/security/jgss/wrapper/GSSNameElement.java:
```diff
@@ -292,9 +292,17 @@ public boolean isAnonymousName() {
+                stub.releaseName(pName);
+            }
+        };
+    }
```
There seem to be a potential optimization here, since you don't even need to create a Cleanable in the case where stub is null or pName is 0? Same for earlier duplication of the same code.

XueleiFan commented:
@wangweij In the following constructor, I'm not sure if the assert right.
```
    NativeGSSContext(long pCtxt, GSSLibStub stub) throws GSSException {
        assert(pContext != 0);
        ...
```

Should it be `assert(pCtxt != 0);`?

dfuch commented on src/java.security.jgss/share/classes/sun/security/jgss/wrapper/NativeGSSContext.java:
```diff
@@ -370,12 +372,36 @@ public void dispose() throws GSSException {
-        if (pContext != 0) {
-            pContext = cStub.deleteContext(pContext);
+
+        if (pContext != 0 && cleanable != null) {
+            cleanable.clean();
             pContext = 0;
```
Here and below: should pContext be set to 0 before calling cleanable.clean() to make sure cleanable.clean() is not called twice? Also should there be some synchronized block?

XueleiFan replied:
> Also should there be some synchronized block?

I thought about the synchronization issue.  It looks like this class is not multiple-threads safe, and thus the upper layer will take care of the synchronization.  So I did not add synchronization in this class update.

> should pContext be set to 0 before calling cleanable.clean() to make sure cleanable.clean() is not called twice? 

Cleanable.clean() will work one time only, no matter how many time it is called.  But it looks like better to make a change here.

dfuch reviewed (approved):
The dispose/cleanable/Cleaner parts look good to me. Please get another reviewer for the security-libs related and native changes.

wangweij commented on test/jdk/sun/security/jgss/GssContextCleanup.java:
```diff
@@ -0,0 +1,64 @@
+        // Check if the object has been collected.
+        if (whm.size() > 0) {
+            throw new RuntimeException("GSSContext object is not released");
+        }
```
I think it's necessary to check the debug message to make sure "[GSSLibStub_deleteContext]" is seen. This test should have already succeeded before your latest update to `NativeGSSContext`.

wangweij replied:
The same for `GSSName`. Where is a test for `GSSCredential`?

XueleiFan replied:
> Where is a test for `GSSCredential`?
I did not find a way to create a GSSCredential object successfully, exception thrown.  Is there an example I can refer to?

wangweij replied:
I think there needs a real Kerberos environment to create a real credential.

wangweij commented:
The latest source change is OK, but I'll look more into it tomorrow.

The test needs some enhancement. Or I can consider contributing one. Sorry I'll be a little busy this week.

XueleiFan commented on src/java.security.jgss/share/classes/sun/security/jgss/wrapper/GSSNameElement.java:
```diff
@@ -48,11 +49,12 @@
 public class GSSNameElement implements GSSNameSpi {
+    private final Cleaner.Cleanable cleanable;
 
     long pName = 0; // Pointer to the gss_name_t structure
```
Did you mean pName? The dispose() method will reset it to zero.  'pName" is used a lot in native implementation.  It may be doable to make it final, but it may be more complicated than I could expect.  I would like to leave it as it is in this PR.

wangweij replied:
IMO, there's no need to reset it to zero in `dispose()`. As for the usage in native implementation, if there is really a case that it gets updated, then the cleanable you registered at the beginning will be useless. Isn't that a real problem?

wangweij replied:
I don't think anything is guaranteed after a name is disposed. I see no problem in doing this.

wangweij commented:
Can you try out this test?

```
diff --git a/test/jdk/sun/security/krb5/auto/Cleaners.java b/test/jdk/sun/security/krb5/auto/Cleaners.java
new file mode 100644
index 00000000000..0eac7c21ee6
--- /dev/null
+++ b/test/jdk/sun/security/krb5/auto/Cleaners.java
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8284490
+ * @summary Remove finalizer method in java.security.jgss
+ * @library /test/lib
+ * @compile -XDignore.symbol.file Cleaners.java
+ * @run main/othervm Cleaners launcher
+ */
+
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.Arrays;
+import java.util.Set;
+
+import jdk.test.lib.Asserts;
+import jdk.test.lib.Platform;
+import jdk.test.lib.process.Proc;
+import org.ietf.jgss.Oid;
+import sun.security.krb5.Config;
+
+public class Cleaners {
+
+    private static final String CONF = "krb5.conf";
+    private static final String KTAB_S = "server.ktab";
+    private static final String KTAB_B = "backend.ktab";
+
+    private static final String HOST = "localhost";
+    private static final String SERVER = "server/" + HOST;
+    private static final String BACKEND = "backend/" + HOST;
+    private static final String USER = "user";
+    private static final char[] PASS = "password".toCharArray();
+    private static final String REALM = "REALM";
+
+    private static final byte[] MSG = "12345678".repeat(128)
+            .getBytes(StandardCharsets.UTF_8);
+
+    public static void main(String[] args) throws Exception {
+
+        Oid oid = new Oid("1.2.840.113554.1.2.2");
+        byte[] token, msg;
+
+        switch (args[0]) {
+            case "launcher" -> {
+                KDC kdc = KDC.create(REALM, HOST, 0, true);
+                kdc.addPrincipal(USER, PASS);
+                kdc.addPrincipalRandKey("krbtgt/" + REALM);
+                kdc.addPrincipalRandKey(SERVER);
+                kdc.addPrincipalRandKey(BACKEND);
+
+                // Native lib might do some name lookup
+                KDC.saveConfig(CONF, kdc,
+                        "dns_lookup_kdc = no",
+                        "ticket_lifetime = 1h",
+                        "dns_lookup_realm = no",
+                        "dns_canonicalize_hostname = false",
+                        "forwardable = true");
+                System.setProperty("java.security.krb5.conf", CONF);
+                Config.refresh();
+
+                // Create kaytab and ccache files for native clients
+                kdc.writeKtab(KTAB_S, false, SERVER);
+                kdc.writeKtab(KTAB_B, false, BACKEND);
+                kdc.kinit(USER, "ccache");
+                if (!Platform.isWindows()) {
+                    Files.setPosixFilePermissions(Paths.get("ccache"),
+                            Set.of(PosixFilePermission.OWNER_READ,
+                                    PosixFilePermission.OWNER_WRITE));
+                }
+                Proc pc = proc("client")
+                        .env("KRB5CCNAME", "FILE:ccache")
+                        .env("KRB5_KTNAME", "none") // Do not try system ktab if ccache fails
+                        .start();
+                Proc ps = proc("server")
+                        .env("KRB5_KTNAME", KTAB_S)
+                        .start();
+                Proc pb = proc("backend")
+                        .env("KRB5_KTNAME", KTAB_B)
+                        .start();
+
+                // Client and server
+                ps.println(pc.readData()); // AP-REQ
+                pc.println(ps.readData()); // AP-REP, mutual auth
+                ps.println(pc.readData()); // wrap msg
+                ps.println(pc.readData()); // mic msg
+
+
+                // Server and backend
+                pb.println(ps.readData()); // AP-REQ
+                ps.println(pb.readData()); // wrap msg
+                ps.println(pb.readData()); // mic msg
+                ensureCleanersCalled(pc);
+                ensureCleanersCalled(ps);
+                ensureCleanersCalled(pb);
+            }
+            case "client" -> {
+                Context c = Context.fromThinAir();
+                c.startAsClient(SERVER, oid);
+                c.x().requestCredDeleg(true);
+                c.x().requestMutualAuth(true);
+                Proc.binOut(c.take(new byte[0])); // AP-REQ
+                c.take(Proc.binIn()); // AP-REP
+                Proc.binOut(c.wrap(MSG, true));
+                Proc.binOut(c.getMic(MSG));
+            }
+            case "server" -> {
+                Context s = Context.fromThinAir();
+                s.startAsServer(oid);
+                token = Proc.binIn(); // AP-REQ
+                Proc.binOut(s.take(token)); // AP-REP
+                msg = s.unwrap(Proc.binIn(), true);
+                Asserts.assertTrue(Arrays.equals(msg, MSG));
+                s.verifyMic(Proc.binIn(), msg);
+                Context s2 = s.delegated();
+                s2.startAsClient(BACKEND, oid);
+                s2.x().requestMutualAuth(false);
+                Proc.binOut(s2.take(new byte[0])); // AP-REQ
+                msg = s2.unwrap(Proc.binIn(), true);
+                Asserts.assertTrue(Arrays.equals(msg, MSG));
+                s2.verifyMic(Proc.binIn(), msg);
+            }
+            case "backend" -> {
+                Context b = Context.fromThinAir();
+                b.startAsServer(oid);
+                token = b.take(Proc.binIn()); // AP-REQ
+                Asserts.assertTrue(token == null);
+                Proc.binOut(b.wrap(MSG, true));
+                Proc.binOut(b.getMic(MSG));
+            }
+        }
+        System.out.println("Prepare for GC");
+        for (int i = 0; i < 10; i++) {
+            System.gc();
+            Thread.sleep(100);
+        }
+    }
+
+    private static void ensureCleanersCalled(Proc p) throws Exception {
+        p.output()
+                .shouldHaveExitValue(0)
+                .stdoutShouldMatch("Prepare for GC(.|\\n)*GSSLibStub_deleteContext")
+                .stdoutShouldMatch("Prepare for GC(.|\\n)*GSSLibStub_releaseName")
+                .stdoutShouldMatch("Prepare for GC(.|\\n)*GSSLibStub_releaseCred");
+    }
+
+    private static Proc proc(String type) throws Exception {
+        return Proc.create("Cleaners")
+                .args(type)
+                .debug(type)
+                .env("KRB5_CONFIG", CONF)
+                .env("KRB5_TRACE", Platform.isWindows() ? "CON" : "/dev/stderr")
+                .prop("sun.security.jgss.native", "true")
+                .prop("javax.security.auth.useSubjectCredsOnly", "false")
+                .prop("sun.security.nativegss.debug", "true");
+    }
+}
diff --git a/test/lib/jdk/test/lib/process/Proc.java b/test/lib/jdk/test/lib/process/Proc.java
index 3541f34144f..574e5761d03 100644
--- a/test/lib/jdk/test/lib/process/Proc.java
+++ b/test/lib/jdk/test/lib/process/Proc.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2013, 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2013, 2022, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -123,6 +123,7 @@ public class Proc {
     private String debug;           // debug flag, controller will show data
                                     // transfer between procs. If debug is set,
                                     // it MUST be different between Procs.
+    private final StringBuilder stdout = new StringBuilder();

     final private static String PREFIX = "PROCISFUN:";

@@ -358,6 +359,9 @@ public class Proc {
     // Reads a line from stdout of proc
     public String readLine() throws IOException {
         String s = br.readLine();
+        if (s != null) {
+            stdout.append(s).append('\n');
+        }
         if (debug != null) {
             System.out.println("PROC: " + debug + " readline: " +
                     (s == null ? "<EOF>" : s));
@@ -402,6 +406,16 @@ public class Proc {
         }
         return p.waitFor();
     }
+
+    // Returns an OutputAnalyzer
+    public OutputAnalyzer output() throws Exception {
+        int exitCode = waitFor();
+        Path stderr = Path.of(getId("stderr"));
+        return new OutputAnalyzer(stdout.toString(),
+                Files.exists(stderr) ? Files.readString(stderr) : "",
+                exitCode);
+    }
+
     // Wait for process end with expected exit code
     public void waitFor(int expected) throws Exception {
         if (p.waitFor() != expected) {
```

wangweij commented:
Sorry, I just updated the test. It should not run on Windows.

XueleiFan commented:
> Can you try out this test?
> 
> ```
> diff --git a/test/jdk/sun/security/krb5/auto/Cleaners.java b/test/jdk/sun/security/krb5/auto/Cleaners.java
> new file mode 100644
> index 00000000000..43f06cb9f60
> --- /dev/null
> +++ b/test/jdk/sun/security/krb5/auto/Cleaners.java
> @@ -0,0 +1,179 @@
> +/*
> + * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
> + * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
> + *
> + * This code is free software; you can redistribute it and/or modify it
> + * under the terms of the GNU General Public License version 2 only, as
> + * published by the Free Software Foundation.
> + *
> + * This code is distributed in the hope that it will be useful, but WITHOUT
> + * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
> + * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
> + * version 2 for more details (a copy is included in the LICENSE file that
> + * accompanied this code).
> + *
> + * You should have received a copy of the GNU General Public License version
> + * 2 along with this work; if not, write to the Free Software Foundation,
> + * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
> + *
> + * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
> + * or visit www.oracle.com if you need additional information or have any
> + * questions.
> + */
> +
> +/*
> + * @test
> + * @bug 8284490
> + * @summary Remove finalizer method in java.security.jgss
> + * @requires os.family != "windows"
> + * @library /test/lib
> + * @compile -XDignore.symbol.file Cleaners.java
> + * @run main/othervm Cleaners launcher
> + */
> +
> +import java.nio.charset.StandardCharsets;
> +import java.nio.file.Files;
> +import java.nio.file.Paths;
> +import java.nio.file.attribute.PosixFilePermission;
> +import java.util.Arrays;
> +import java.util.Set;
> +
> +import jdk.test.lib.Asserts;
> +import jdk.test.lib.process.Proc;
> +import org.ietf.jgss.Oid;
> +import sun.security.krb5.Config;
> +
> +public class Cleaners {
> +
> +    private static final String CONF = "krb5.conf";
> +    private static final String KTAB_S = "server.ktab";
> +    private static final String KTAB_B = "backend.ktab";
> +
> +    private static final String HOST = "localhost";
> +    private static final String SERVER = "server/" + HOST;
> +    private static final String BACKEND = "backend/" + HOST;
> +    private static final String USER = "user";
> +    private static final char[] PASS = "password".toCharArray();
> +    private static final String REALM = "REALM";
> +
> +    private static final byte[] MSG = "12345678".repeat(128)
> +            .getBytes(StandardCharsets.UTF_8);
> +
> +    public static void main(String[] args) throws Exception {
> +
> +        Oid oid = new Oid("1.2.840.113554.1.2.2");
> +        byte[] token, msg;
> +
> +        switch (args[0]) {
> +            case "launcher" -> {
> +                KDC kdc = KDC.create(REALM, HOST, 0, true);
> +                kdc.addPrincipal(USER, PASS);
> +                kdc.addPrincipalRandKey("krbtgt/" + REALM);
> +                kdc.addPrincipalRandKey(SERVER);
> +                kdc.addPrincipalRandKey(BACKEND);
> +
> +                // Native lib might do some name lookup
> +                KDC.saveConfig(CONF, kdc,
> +                        "dns_lookup_kdc = no",
> +                        "ticket_lifetime = 1h",
> +                        "dns_lookup_realm = no",
> +                        "dns_canonicalize_hostname = false",
> +                        "forwardable = true");
> +                System.setProperty("java.security.krb5.conf", CONF);
> +                Config.refresh();
> +
> +                // Create kaytab and ccache files for native clients
> +                kdc.writeKtab(KTAB_S, false, SERVER);
> +                kdc.writeKtab(KTAB_B, false, BACKEND);
> +                kdc.kinit(USER, "ccache");
> +                Files.setPosixFilePermissions(Paths.get("ccache"),
> +                        Set.of(PosixFilePermission.OWNER_READ,
> +                                PosixFilePermission.OWNER_WRITE));
> +
> +                Proc pc = proc("client")
> +                        .env("KRB5CCNAME", "FILE:ccache")
> +                        .env("KRB5_KTNAME", "none") // Do not try system ktab if ccache fails
> +                        .start();
> +                Proc ps = proc("server")
> +                        .env("KRB5_KTNAME", KTAB_S)
> +                        .start();
> +                Proc pb = proc("backend")
> +                        .env("KRB5_KTNAME", KTAB_B)
> +                        .start();
> +
> +                // Client and server
> +                ps.println(pc.readData()); // AP-REQ
> +                pc.println(ps.readData()); // AP-REP, mutual auth
> +                ps.println(pc.readData()); // wrap msg
> +                ps.println(pc.readData()); // mic msg
> +
> +                // Server and backend
> +                pb.println(ps.readData()); // AP-REQ
> +                ps.println(pb.readData()); // wrap msg
> +                ps.println(pb.readData()); // mic msg
> +
> +                ensureCleanersCalled(pc);
> +                ensureCleanersCalled(ps);
> +                ensureCleanersCalled(pb);
> +            }
> +            case "client" -> {
> +                Context c = Context.fromThinAir();
> +                c.startAsClient(SERVER, oid);
> +                c.x().requestCredDeleg(true);
> +                c.x().requestMutualAuth(true);
> +                Proc.binOut(c.take(new byte[0])); // AP-REQ
> +                c.take(Proc.binIn()); // AP-REP
> +                Proc.binOut(c.wrap(MSG, true));
> +                Proc.binOut(c.getMic(MSG));
> +            }
> +            case "server" -> {
> +                Context s = Context.fromThinAir();
> +                s.startAsServer(oid);
> +                token = Proc.binIn(); // AP-REQ
> +                Proc.binOut(s.take(token)); // AP-REP
> +                msg = s.unwrap(Proc.binIn(), true);
> +                Asserts.assertTrue(Arrays.equals(msg, MSG));
> +                s.verifyMic(Proc.binIn(), msg);
> +                Context s2 = s.delegated();
> +                s2.startAsClient(BACKEND, oid);
> +                s2.x().requestMutualAuth(false);
> +                Proc.binOut(s2.take(new byte[0])); // AP-REQ
> +                msg = s2.unwrap(Proc.binIn(), true);
> +                Asserts.assertTrue(Arrays.equals(msg, MSG));
> +                s2.verifyMic(Proc.binIn(), msg);
> +            }
> +            case "backend" -> {
> +                Context b = Context.fromThinAir();
> +                b.startAsServer(oid);
> +                token = b.take(Proc.binIn()); // AP-REQ
> +                Asserts.assertTrue(token == null);
> +                Proc.binOut(b.wrap(MSG, true));
> +                Proc.binOut(b.getMic(MSG));
> +            }
> +        }
> +        System.out.println("Prepare for GC");
> +        for (int i = 0; i < 10; i++) {
> +            System.gc();
> +            Thread.sleep(100);
> +        }
> +    }
> +
> +    private static void ensureCleanersCalled(Proc p) throws Exception {
> +        p.output()
> +                .shouldHaveExitValue(0)
> +                .stdoutShouldMatch("Prepare for GC(.|\\n)*GSSLibStub_deleteContext")
> +                .stdoutShouldMatch("Prepare for GC(.|\\n)*GSSLibStub_releaseName")
> +                .stdoutShouldMatch("Prepare for GC(.|\\n)*GSSLibStub_releaseCred");
> +    }
> +
> +    private static Proc proc(String type) throws Exception {
> +        return Proc.create("Cleaners")
> +                .args(type)
> +                .debug(type)
> +                .env("KRB5_CONFIG", CONF)
> +                .env("KRB5_TRACE", "/dev/stderr")
> +                .prop("sun.security.jgss.native", "true")
> +                .prop("javax.security.auth.useSubjectCredsOnly", "false")
> +                .prop("sun.security.nativegss.debug", "true");
> +    }
> +}
> diff --git a/test/lib/jdk/test/lib/process/Proc.java b/test/lib/jdk/test/lib/process/Proc.java
> index 3541f34144f..574e5761d03 100644
> --- a/test/lib/jdk/test/lib/process/Proc.java
> +++ b/test/lib/jdk/test/lib/process/Proc.java
> @@ -1,5 +1,5 @@
>  /*
> - * Copyright (c) 2013, 2021, Oracle and/or its affiliates. All rights reserved.
> + * Copyright (c) 2013, 2022, Oracle and/or its affiliates. All rights reserved.
>   * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
>   *
>   * This code is free software; you can redistribute it and/or modify it
> @@ -123,6 +123,7 @@ public class Proc {
>      private String debug;           // debug flag, controller will show data
>                                      // transfer between procs. If debug is set,
>                                      // it MUST be different between Procs.
> +    private final StringBuilder stdout = new StringBuilder();
> 
>      final private static String PREFIX = "PROCISFUN:";
> 
> @@ -358,6 +359,9 @@ public class Proc {
>      // Reads a line from stdout of proc
>      public String readLine() throws IOException {
>          String s = br.readLine();
> +        if (s != null) {
> +            stdout.append(s).append('\n');
> +        }
>          if (debug != null) {
>              System.out.println("PROC: " + debug + " readline: " +
>                      (s == null ? "<EOF>" : s));
> @@ -402,6 +406,16 @@ public class Proc {
>          }
>          return p.waitFor();
>      }
> +
> +    // Returns an OutputAnalyzer
> +    public OutputAnalyzer output() throws Exception {
> +        int exitCode = waitFor();
> +        Path stderr = Path.of(getId("stderr"));
> +        return new OutputAnalyzer(stdout.toString(),
> +                Files.exists(stderr) ? Files.readString(stderr) : "",
> +                exitCode);
> +    }
> +
>      // Wait for process end with expected exit code
>      public void waitFor(int expected) throws Exception {
>          if (p.waitFor() != expected) {
> ```

wangweij commented:
I see you are still using the 1st version of the `Cleaners.java` test which runs on Windows. Please update to the current version (I've updated the code in place).

XueleiFan commented:
> I see you are still using the 1st version of the `Cleaners.java` test which runs on Windows. Please update to the current version (I've updated the code in place).

Oops, I missed it.  Updated.

wangweij commented:
Oops, one test failed. The test output shows native bridge is not used. I believe this is because the test is not run with `othervm` but the test is setting system properties. Please change them to use `othervm`.

XueleiFan commented:
>  please change them to use `othervm`.

Thanks for the catch. Updated to use othervm.

wangweij commented:
I see you removed the `Thread.sleep(100)` calls. Given the failure of another similar test, maybe it's safer to add them back?

XueleiFan commented:
> I see you removed the `Thread.sleep(100)` calls. Given the failure of another similar test, maybe it's safer to add them back?

Yes.  I'm evaluating if other proposal works or not.  Otherwise, I will add the sleep back.

wangweij commented:
I'm sure you have well established testing environment as well, but if you need any help to test out your "other proposals", I'm glad to help.

XueleiFan commented:
The Thread.sleep() was added back.  Without the sleep, the test may fail intermittent.

wangweij commented:
My understanding is such kind of tests that rely on GC is fundementally unreliable. I would preemptively add an intermittent keyword to it. :-)

XueleiFan commented:
/clean

wangweij reviewed (approved):
LGTM

bchristi-git commented:
Hi. Sorry, I should have brought this up earlier, but there is a jtreg test library to help with ensuring the GC runs, 
`test/lib/jdk/test/lib/util/ForceGC.java`. You might consider replacing the test code that runs/checks the GC with `ForceGC.await()`.

XueleiFan commented:
> Hi. Sorry, I should have brought this up earlier, but there is a jtreg test library to help with ensuring the GC runs, `test/lib/jdk/test/lib/util/ForceGC.java`. You might consider replacing the test code that runs/checks the GC with `ForceGC.await()`.

It looks like a graceful GC utility.  There are also some other test cases I committed that do no use ForceGC.  I filed a [new bug](https://bugs.openjdk.java.net/browse/JDK-8286045), and will make the update all together.  Thank you.

XueleiFan commented:
/integrate

seanjmullan commented:
@XueleiFan Did you check to make sure someone ran Mach5 and the results were ok before integrating?

XueleiFan commented:
> @XueleiFan Did you check to make sure someone ran Mach5 and the results were ok before integrating?

If I'm right, Weijun made the test.  Please let me know if there are mach5 testing failures.

BTW, it would be nice if Mach5 testing could be supported to run in Github.
